coins/monero/src/wallet/seed/mod.rs

```diff
@@ -67,26 +67,52 @@ impl Seed {
       CLASSIC_SEED_LENGTH | CLASSIC_SEED_LENGTH_WITH_CHECKSUM => {
         ClassicSeed::from_string(words).map(Seed::Classic)
       }
+      POLYSEED_LENGTH => PolySeed::from_string(words).map(Seed::PolySeed),
       _ => Err(SeedError::InvalidSeedLength)?,
     }
   }
 
   /// Create a Seed from entropy.
-  pub fn from_entropy(lang: Language, entropy: Zeroizing<[u8; 32]>) -> Option<Seed> {
-    ClassicSeed::from_entropy(lang, entropy).map(Seed::Classic)
+  pub fn from_entropy(seed_type: SeedType, entropy: Zeroizing<[u8; 32]>) -> Option<Seed> {
+    match seed_type {
+      SeedType::Classic(lang) => ClassicSeed::from_entropy(lang, entropy).map(Seed::Classic),
+      SeedType::PolySeed(lang) => {
+        let birthday = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs();
+        let data = PolySeedData { birthday, features: 0, entropy, checksum: 0 };
+        PolySeed::from_polyseed_data(data, lang).map(Seed::PolySeed)
+      }
+    }
+  }
+
+  /// Create a Seed from given data.
+  pub fn from_polyseed_data(data: PolySeedData, lang: polyseed::Language) -> Option<Seed> {
+    PolySeed::from_polyseed_data(data, lang).map(Seed::PolySeed)
   }
 
   /// Convert a seed to a String.
   pub fn to_string(&self) -> Zeroizing<String> {
     match self {
       Seed::Classic(seed) => seed.to_string(),
+      Seed::PolySeed(seed) => seed.to_string(),
     }
   }
 
   /// Return the entropy for this seed.
   pub fn entropy(&self) -> Zeroizing<[u8; 32]> {
     match self {
       Seed::Classic(seed) => seed.entropy(),
+      Seed::PolySeed(seed) => seed.polyseed_data().entropy,
+    }
+  }
+
+  /// Return the polyseed data for this seed.
+  /// Return 0 values if it is a classic seed.
```

Comment:
This should return an Option, not 0.

I'd also rather not have "polyseed_data", yet a dedicated birthday/features function. The birthday for a Classic seed should be 0 (not because it's unset but because no other value is safe). For features, it'd likely also be 0, signifying no features? Yet we may want to explicitly handle features in our API instead of relying on a uint expected to be interpreted per Polyseed. This may mean features returns Vec<Feature> or similar.

Reply:
the problem with having separate birthday/features function is that it is inefficient. We will need to re-decode the seed per call. In one scenario, user needs all polyseed data and makes 1 call to `polyseed_data`, in another user makes 3 call to get the data they want and we will re-decode the same seed 3 times, although all the data was already available after the first call.